### ISSUE TITLE
[bug] “full worker queue” and cannot post/favourite/… any more

### ISSUE THREAD

**Issue by mirabilos** [superseriousbusiness/gotosocial#1559]
### Describe the bug with a clear and concise description of what the bug is.

Once is happenstance, twice is enemy (or bug) action.

I try to post, and “nothing happens”; the toot button in Pinafore spins round and eventually returns to a clickable button without clearing the input field, indicating the post didn’t go through.

### What's your GoToSocial Version?

0.7.1

### GoToSocial Arch

amd64 binary

### What happened?

I saw this in the log…

```
@4000000063fb05fb0e3fa2ec timestamp="26/02/2023 07:10:41.238" func=concurrency.(*WorkerPool).Queue level=WARN type=worker.Worker[messages.FromClientAPI] queue=400 msg="full worker queue"
```

… and Pinafore just spins endlessly instead of posting.

### What you expected to happen?

Things to work ;-)

### How to reproduce it?

This happened the second time tonight, so I expect this to happen under some load eventually?

Incidentally, I had not been using my instance for a couple of hours, so all activity in the meantime was network stuff coming in.

### Anything else we need to know?

In that state, shutting down (and restarting) does not work cleanly either:

```
@4000000063fb06d204514344 timestamp="26/02/2023 07:14:16.072" func=server.glob..func1 level=INFO msg="received signal terminated, shutting down"
@4000000063fb06d204514efc timestamp="26/02/2023 07:14:16.072" func=router.(*router).Stop level=INFO msg="shutting down http router with 30s grace period"
@4000000063fb06f00a480ea4 2023/02/26 07:14:46 error executing command: error closing gotosocial service: error shutting down http router: context deadline exceeded
@4000000063fb06f0119555cc starting @ Sun Feb 26 07:14:46 UTC 2023
[…]
```

That last line is my dæmontools start wrapper, so it did terminate after a while.

Hitting the Toot button again after the restart makes things go through immediately.

Looking at the post I was replying to, I see multiple copies of the message in my local Pinafore. Looking at the foreign user’s instance, I only see one (the last) copy. Might be a local cache in Pinafore or one in GtS.

**Comment by mirabilos** [comment]
First occurrence ~six hours ago: https://toot.mirbsd.org/@mirabilos/statuses/01GT5JRRFTW2RDZNBR20Q2X2R7

**Comment by NyaaaWhatsUpDoc** [comment]
Where there many entries in the log regarding full worker queue? (this would imply a lock being held somewhere that prevents a worker being freed).

Alternatively, do you have request throttling enabled? If not then what are your server's specs?

**Comment by mirabilos** [comment]
kim dixit:

>Where there many entries in the log regarding full worker queue? (this
>would imply a lock being held somewhere that prevents a worker being
>freed).

Depends on your definition of “many”. Like you, I’m using a rotating
thing for the logs to not overflow; while this was happening, I saw
some of them pass by (especially after I did something (post/fav/…).

>Alternatively, do you have request throttling enabled?

I do, with what I think are the default settings:

advanced-rate-limit-requests: 300
advanced-throttling-multiplier: 8
advanced-throttling-retry-after: "45s"

Does that apply to local actions as well? In a manner that remote
actions fill up the queue for local actions even?

>If not then what are your server's specs? 

I rent it as VM; the current CPU is:

$ cat /proc/cpuinfo
processor       : 0
vendor_id       : AuthenticAMD
cpu family      : 6
model           : 42
model name      : Intel Xeon E312xx (Sandy Bridge)
stepping        : 1
microcode       : 0x1000065
cpu MHz         : 2600.000
cache size      : 512 KB
physical id     : 0
siblings        : 1
core id         : 0
cpu cores       : 1
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 13
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 syscall nx rdtscp lm nopl cpuid tsc_known_freq pni pclmulqdq ssse3 cx16 sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx hypervisor lahf_lm 3dnowprefetch vmmcall xsaveopt arat
bugs            : fxsave_leak sysret_ss_attrs null_seg spectre_v1 spectre_v2 spec_store_bypass
bogomips        : 5200.00
TLB size        : 1024 4K pages
clflush size    : 64
cache_alignment : 64
address sizes   : 40 bits physical, 48 bits virtual
power management:

I’m apparently guaranteed one full vCPU (no CPU overcommitting).

$ free
               total        used        free      shared  buff/cache   available
Mem:         1000028      197276       94452      147548      708300      500912
Swap:        2097148       66600     2030548

The system hosts a not-too-busy Postfix, a probably-more-busy
Apache 2 + PHP with a not-busy legacy site, GtS, and a PostgreSQL
cluster shared between that PHP site and GtS; fail2ban is active.
Oh and apparently another old thing that serves a static text file
and receives a response back via CGI.

Apache served 36394 requests on 25 Feb:

• 46 attacks with no Host header and no destination IP(?!), 105 on the IP
• 819 “internal dummy connection” from mod_ssl(I think?)
• 6169 static+CGI calls
• 20729 calls to the GtS vhost
• 5878 old site static content
• 1152 other misc. static/404 requests
• 1496 remaining, probably 99% old site PHP

I do have MaxRequestWorkers 200 set, 400 was too much for the machine.

AFAICT the issue started after upgrading from 0.6 so it may be related
to request trottling.

bye,
//mirabilos
-- 
(gnutls can also be used, but if you are compiling lynx for your own use,
there is no reason to consider using that package)
	-- Thomas E. Dickey on the Lynx mailing list, about OpenSSL
